integration/experiment/power_sweep/gen_policy_recommendation.py

```diff
@@ -0,0 +1,350 @@
+        if rtmodel:
+            best_runtime = rtmodel.test(best_pl)
+        else:
+            best_runtime = None
+        return {'power': best_pl,
+                'runtime': best_runtime,
+                'energy': best_energy}
+    else:
+        rt_predictions = rtmodel.batch_test(plrange)
+        rt_at_tdp = rtmodel.test(pltdp)
+        constrained_values = [
+                (energy, pl, runtime)
+                for pl, runtime, energy
+                in zip(plrange, rt_predictions, en_predictions)
+                if runtime <= (1 + max_degradation) * rt_at_tdp]
+        best_energy, best_pl, best_runtime = min(constrained_values)
+        return {'power': best_pl,
+                'runtime': best_runtime,
+                'energy': best_energy}
+
+
+def main(full_df,
+        region_filter,
+        dump_prefix,
+        min_pl,
+        max_pl,
+        tdp,
+        min_energy,
+        max_degradation):
+    """
+    The main function. full_df is a report collection dataframe,
+    region_filter is a list of regions to include, dump_prefix
+    a filename prefix for debugging output (if specified), min_pl,
+    max_pl, tdp are the minimum, maximum, and reference power limits,
+    respectively, min_energy is a flag indicating whether or not
+    minimum energy is sought not subject to a constraint on runtime
+    (True if runtime should be ignored) and max_degradation specifies
+    what maximum runtime degradation is accepted."""
+    df = extract_columns(full_df, args.region_filter)
+    if dump_prefix:
+        df.to_csv("{}.dat".format(dump_prefix))
+        dump_stats_summary(df, "{}.stats".format(dump_prefix))
+
+    runtime_model = CubicPowerModel()
+    energy_model = CubicPowerModel()
+
+    runtime_model.train(df, key='runtime')
+    energy_model.train(df, key='energy')
+
+    plrange = [min_pl + i for i in range(int(max_pl - min_pl))] + [max_pl]
+    best_policy = policy_min_energy(plrange, energy_model, runtime_model, tdp,
+            max_degradation=(None if args.min_energy else args.max_degradation))
+    tdprt, tdpen = runtime_model.test(tdp), energy_model.test(tdp)
+
+    return {'tdp': {'power': tdp, 'runtime': tdprt, 'energy': tdpen},
+            'best': best_policy}
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    common_args.add_min_power(parser)
+    common_args.add_max_power(parser)
+    parser.add_argument('--path', required=True,
+                        help='path containing reports and machine.json')
+    parser.add_argument('--region_filter', default=None,
+                        help='comma-separated list of regions to include, '
+                             'default to include all regions')
+    parser.add_argument('--dump_prefix',
+                        help='prefix to dump statistics to, empty to not dump '
+                             'stats')
+    parser.add_argument('--tdp', default=-1, type=float,
```

Comment:
can this just use None for default, and check for None on 304 below?  any other place we can get negative values?

Reply:
Done.